flaky TestTransportReplaceWithClosed

```
--- FAIL: TestTransportReplaceWithClosed (0.03s)
    --- FAIL: TestTransportReplaceWithClosed/local (0.03s)
        transport_test.go:656: 
            	Error Trace:	/home/parallels/src/go/src/github.com/quic-go/quic-go/transport_test.go:656
            	            				/home/parallels/src/go/src/github.com/quic-go/quic-go/transport_test.go:604
            	Error:      	Not equal: 
            	            	expected: []byte{0x66, 0x6f, 0x6f, 0x62, 0x61, 0x72}
            	            	actual  : []byte{0x52, 0x4e, 0xf8, 0xda, 0x93, 0xbc, 0x43, 0xe3, 0x5c, 0xda, 0x58, 0x68, 0x30, 0xb0, 0xea, 0x13, 0xcc, 0xdb, 0xfd, 0xa1, 0x3d, 0xf8, 0x48, 0x20, 0xff, 0x64, 0x15, 0x40, 0x57, 0x6e, 0x1a, 0x8b, 0x7f, 0x3e, 0x7a, 0x4d, 0x8a, 0x35, 0xf6, 0xe0, 0x9, 0x23}
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,3 +1,5 @@
            	            	-([]uint8) (len=6) {
            	            	- 00000000  66 6f 6f 62 61 72                                 |foobar|
            	            	+([]uint8) (len=42) {
            	            	+ 00000000  52 4e f8 da 93 bc 43 e3  5c da 58 68 30 b0 ea 13  |RN....C.\.Xh0...|
            	            	+ 00000010  cc db fd a1 3d f8 48 20  ff 64 15 40 57 6e 1a 8b  |....=.H .d.@Wn..|
            	            	+ 00000020  7f 3e 7a 4d 8a 35 f6 e0  09 23                    |.>zM.5...#|
            	            	 }
            	Test:       	TestTransportReplaceWithClosed/local

FAIL
exit status 1

--- a/transport_test.go
+++ b/transport_test.go
@@ -5,8 +5,10 @@ import (
 	"context"
 	"crypto/tls"
 	"errors"
+	"math"
 	"net"
-	"os"
+	"runtime"
+	"sync/atomic"
 	"syscall"
 	"testing"
 	"time"
@@ -598,6 +600,8 @@ func TestTransportDialingVersionNegotiation(t *testing.T) {
 }
 
 func TestTransportReplaceWithClosed(t *testing.T) {
+	t.Setenv("QUIC_GO_DISABLE_RECEIVE_BUFFER_WARNING", "true")
+
 	t.Run("local", func(t *testing.T) {
 		testTransportReplaceWithClosed(t, true)
 	})
@@ -616,7 +620,7 @@ func testTransportReplaceWithClosed(t *testing.T, local bool) {
 	tr.init(true)
 	defer tr.Close()
 
-	dur := scaleDuration(10 * time.Millisecond)
+	dur := scaleDuration(20 * time.Millisecond)
 
 	var closePacket []byte
 	if local {
@@ -627,7 +631,6 @@ func testTransportReplaceWithClosed(t *testing.T, local bool) {
 	connID := protocol.ParseConnectionID([]byte{4, 3, 2, 1})
 	m := (*packetHandlerMap)(tr)
 	require.True(t, m.Add(connID, handler))
-	start := time.Now()
 	m.ReplaceWithClosed([]protocol.ConnectionID{connID}, closePacket, dur)
 
 	p := make([]byte, 100)
@@ -635,38 +638,64 @@ func testTransportReplaceWithClosed(t *testing.T, local bool) {
 	copy(p[1:], connID.Bytes())
 
 	conn := newUDPConnLocalhost(t)
-	var sent int
-	for now := range time.NewTicker(dur / 20).C {
-		_, err := conn.WriteTo(p, tr.Conn.LocalAddr())
-		require.NoError(t, err)
-		sent++
-		if now.After(start.Add(dur / 2)) {
-			break
+	var sent atomic.Int64
+	errChan := make(chan error, 1)
+	stopSending := make(chan struct{})
+	go func() {
+		defer close(errChan)
+		ticker := time.NewTicker(dur / 50)
+		timeout := time.NewTimer(scaleDuration(time.Second))
+		for {
+			select {
+			case <-stopSending:
+				return
+			case <-timeout.C:
+				errChan <- errors.New("timeout")
+				return
+			case <-ticker.C:
+			}
+			if _, err := conn.WriteTo(p, tr.Conn.LocalAddr()); err != nil {
+				errChan <- err
+				return
+			}
+			sent.Add(1)
 		}
-	}
+	}()
+
 	// For locally closed connections, CONNECTION_CLOSE packets are sent with an exponential backoff
-	for i := 0; i*i < sent; i++ {
-		conn.SetReadDeadline(time.Now().Add(time.Second))
-		b := make([]byte, 100)
-		if local {
-			n, _, err := conn.ReadFrom(b)
-			require.NoError(t, err)
-			require.Equal(t, []byte("foobar"), b[:n])
-		}
-	}
-	// Afterwards, we receive a stateless reset, not a copy of the CONNECTION_CLOSE packet.
-	// Retry a few times, since the connection is deleted from the map on a timer.
-	require.Eventually(t, func() bool {
-		_, err := conn.WriteTo(p, tr.Conn.LocalAddr())
-		require.NoError(t, err)
-		conn.SetReadDeadline(time.Now().Add(dur / 4))
+	var received int
+	conn.SetReadDeadline(time.Now().Add(scaleDuration(time.Second)))
+	for {
 		b := make([]byte, 100)
 		n, _, err := conn.ReadFrom(b)
-		if errors.Is(err, os.ErrDeadlineExceeded) || bytes.Equal(b[:n], []byte("foobar")) {
-			return false
-		}
 		require.NoError(t, err)
-		require.GreaterOrEqual(t, n, protocol.MinStatelessResetSize)
-		return true
-	}, scaleDuration(200*time.Millisecond), scaleDuration(10*time.Millisecond))
+		// at some point, the connection is cleaned up, and we'll receive a stateless reset
+		if !bytes.Equal(b[:n], []byte("foobar")) {
+			require.GreaterOrEqual(t, n, protocol.MinStatelessResetSize)
+			close(stopSending) // stop sending packets
+			break
+		}
+		received++
+	}
+
+	select {
+	case err := <-errChan:
+		require.NoError(t, err)
+	case <-time.After(time.Second):
+		t.Fatal("timeout")
+	}
+
+	numSent := sent.Load()
+	if !local {
+		require.Zero(t, received)
+		t.Logf("sent %d packets", numSent)
+		return
+	}
+	t.Logf("sent %d packets, received %d CONNECTION_CLOSE copies", numSent, received)
+	// timer resolution on Windows is terrible
+	if runtime.GOOS != "windows" {
+		require.GreaterOrEqual(t, numSent, int64(8))
+	}
+	require.GreaterOrEqual(t, received, int(math.Floor(math.Log2(float64(numSent)))))
+	require.LessOrEqual(t, received, int(math.Ceil(math.Log2(float64(numSent)))))
 }

